girder/models/file.py

```diff
@@ -93,19 +93,44 @@ def download(self, file, offset=0, headers=True, endByte=None,
         :type contentDisposition: str or None
         :type extraParameters: str or None
         """
+        events.trigger('model.file.download.request', info={
+            'file': file, 'startByte': offset, 'endByte': endByte})
+
         if file.get('assetstoreId'):
-            return self.getAssetstoreAdapter(file).downloadFile(
-                file, offset=offset, headers=headers, endByte=endByte,
-                contentDisposition=contentDisposition,
-                extraParameters=extraParameters)
+            try:
+                fileDownload = self.getAssetstoreAdapter(file).downloadFile(
+                    file, offset=offset, headers=headers, endByte=endByte,
+                    contentDisposition=contentDisposition,
+                    extraParameters=extraParameters)
+
+                def downloadGenerator():
+                    for data in fileDownload():
+                        yield data
+                    if endByte is None or endByte >= file['size']:
+                        events.trigger('model.file.download.complete', info={
+                            'file': file, 'startByte': offset,
+                            'endByte': endByte, 'redirect': False})
+                return downloadGenerator
+            except cherrypy.HTTPRedirect:
+                events.trigger('model.file.download.complete', info={
+                    'file': file, 'startByte': offset, 'endByte':
+                    endByte, 'redirect': True})
+                raise
         elif file.get('linkUrl'):
             if headers:
+                events.trigger('model.file.download.complete', info={
+                    'file': file, 'startByte': offset, 'endByte':
+                    endByte, 'redirect': True})
                 raise cherrypy.HTTPRedirect(file['linkUrl'])
             else:
                 endByte = endByte or len(file['linkUrl'])
 
                 def stream():
                     yield file['linkUrl'][offset:endByte]
+                    if endByte >= len(file['linkUrl']):
+                        events.trigger('model.file.download.complete', info={
+                            'file': file, 'startByte': offset, 'endByte':
```

Comment:
Guess we don't have a style rule to enforce this, but could we write this dict with one key per line?